File: Python3/demo/demo_ZNDCJ.py
"""
Shock and Detonation Toolbox Demo Program

Generate plots and output files for a ZND detonation with the shock front
traveling at the CJ speed.

################################################################################
Theory, numerical methods and applications are described in the following report:

SDToolbox - Numerical Tools for Shock and Detonation Wave Modeling,
Explosion Dynamics Laboratory, Contributors: S. Browne, J. Ziegler,
N. Bitter, B. Schmidt, J. Lawson and J. E. Shepherd, GALCIT
Technical Report FM2018.001 Revised January 2021.

Please cite this report and the website if you use these routines. 

Please refer to LICENCE.txt or the above report for copyright and disclaimers.

http://shepherd.caltech.edu/EDL/PublicResources/sdt/

################################################################################ 
Updated September 2018
Tested with: 
    Python 3.5 and 3.6, Cantera 2.3 and 2.4
Under these operating systems:
    Windows 8.1, Windows 10, Linux (Debian 9)
"""

from sdtoolbox.postshock import CJspeed, PostShock_fr
from sdtoolbox.znd import zndsolve
from sdtoolbox.utilities import CJspeed_plot, znd_plot, znd_fileout
import cantera as ct

P1 = 20000. 
T1 = 300.
q = 'CH4:1 O2:2 N2:1.25'
mech = 'gri30_highT.yaml'
file_name = 'CH4_air'


# Find CJ speed and related data, make CJ diagnostic plots
cj_speed,R2,plot_data = CJspeed(P1,T1,q,mech,fullOutput=True)
CJspeed_plot(plot_data,cj_speed)

# Set up gas object
gas1 = ct.Solution(mech)
gas1.TPX = T1,P1,q

# Find post shock state for given speed, start slightly above CJ for this case to avoid
# getting stuck at sonic point
U = 1.005*cj_speed
gas = PostShock_fr(U, P1, T1, q, mech)

"""
CAUTIONS:
If you have trouble getting a converged solution with the ode solver, this is usually associated with
large mechanisms for hydrocarbons. There are often species that are present in very small
amounts at the end of the reaction zone and change rapidly in the energy release portion of the
reaction zone. Although these are not important for obtaining a solution, if the solver takes too large a
time step, negative species amounts will result in an error.  Implicit solvers that automatically adjust the time step
can run into trouble if there is a sudden change in conditions, which happens within energy release zone for compositions and conditions with long induction zone and short energy release zone. 


There are three approaches to dealing with these problems.

1.Switch solvers.
a. For python,  use "LSODA" or "BDF", these are more robust alternatives to the 'Radau' solver that was originally used.  A method
parameter has been added to the calls and the default is being changed to 'LSODA'.
b.  For Matlab,  try "ode23tb" instead of the "ode15s" that is the default.  However, it is often necessary to reduce the max_step
and tolerances for Matlab.

2. Reduce the tolerance paramters, "absTol", "relTol"

3.  Reduce the "max_step" parameter   

Examine the species (particularly the minor species) near the energy
release region to  determine what sort of abs_tol and max_step are needed.  The values can be surprisingly small in order to
avoid oscillations in species concentrations.  

"""

# Solve ZND ODEs, make ZND plots
znd_out = zndsolve(gas,gas1,U,relTol=1e-7,absTol=1e-8,t_end=5e-5, advanced_output=True,Method='LSODA')
znd_plot(znd_out,xscale='log',maxx=10E-3,major_species=['CH4', 'O2', 'H2O', 'CO2'],minor_species=['H', 'O', 'OH', 'CH2O','CO', 'H2'])
znd_fileout(file_name,znd_out)

print('Reaction zone pulse width (exothermic length) = %.4g m' % znd_out['exo_len_ZND'])
print('Reaction zone induction length = %.4g m' % znd_out['ind_len_ZND'])
print('Reaction zone pulse time (exothermic time) = %.4g s' % znd_out['exo_time_ZND'])
print('Reaction zone induction time = %.4g s' % znd_out['ind_time_ZND'])

File: MATLAB/Demo/demo_ZNDCJ.m
% Shock and Detonation Toolbox Demo Program
% 
% Generate plots and output files for a ZND detonation with the shock front
% traveling at the CJ speed.
%  
% ################################################################################
% Theory, numerical methods and applications are described in the following report:
% 
% SDToolbox:  Numerical Tools for Shock and Detonation Wave Modeling.  
% S. T. Kao, J. L. Zeigler, N. P. Bitter, B. E. Schmidt, J. M. Lawson, and J. E. Shepherd. 
% GALCIT Report FM2018.001.  California Institute of Technology, 2023. 
% https://shepherd.caltech.edu/EDL/PublicResources/sdt/doc/ShockDetonation/ShockDetonation.pdf
% 
% Please cite this report and the website: http://shepherd.caltech.edu/EDL/PublicResources/sdt/ 
% if you use these routines. Refer to LICENCE.txt or the above report for copyright and disclaimers.
% ################################################################################ 
% Updated April 2026
% Requires Cantera 3.0, Downloads and documentation at https://cantera.org/3.0/sphinx/html/matlab/index.html
% Tested with Matlab R2024b and Win11
%%
clear;clc;close all;
disp('demo_ZNDCJ')
file_name = 'CH4_air';
P1 = 20E3; T1 = 295; 
q = 'CH4:1 O2:2 N2:1.25';
mech = 'gri30_highT.yaml';
[cj_speed, curve, ~, dnew, plot_data] = CJspeed(P1, T1, q, mech);
%CJspeed_plot(2,plot_data,curve,dnew)

gas1 = Solution(mech);
set(gas1, 'T', T1, 'P', P1, 'X', q);
U_s = 1.005*cj_speed;  % necessary start slightly overdriven not to get stuck at sonic point
% FIND POST SHOCK STATE FOR GIVEN SPEED
[gas] = PostShock_fr(U_s, P1, T1, q, mech);

% SOLVE ZND DETONATION ODES
[out] = zndsolve(gas,gas1,U_s,'advanced_output',true,'t_end',5E-5);
znd_plot(out,'xscale','log','xunit','mm','xmult',1E3,'species_min',1E-8,'maxx',10,'major_species',{'CH4', 'O2', 'H2O', 'CO2'},...
    'minor_species',{'H', 'O', 'OH', 'CH2O','CO', 'H2'},'species_plt','mole','prnfig',false);
znd_fileout(file_name,out)
disp(['Mech ',mech]);
disp(['mix ',q])
disp(['CJ speed = ',num2str(cj_speed),' m/s']);
disp(['Reaction zone pulse width (exothermic length) = ',num2str(out.exo_len_ZND),' m']);
disp(['Reaction zone induction length = ',num2str(out.ind_len_ZND),' m']);
disp(['Reaction zone pulse time (exothermic time) = ',num2str(out.exo_time_ZND),' s']);
disp(['Reaction zone induction time = ',num2str(out.ind_time_ZND),' s']);
